### PR TITLE
[notification] namespaced channel type

### PR DESCRIPTION
**Pull request by qnkhuat**
This PR requires all API to have `channel.type` to be namespaced under `:channel`

The APIs were introduced in https://github.com/metabase/metabase/pull/45207 and it was accepting un-namespaced channel type. `E.g: slack, http` hence the `channel.type` is storing un-namespaced keyword.

Since all channel apis now work with namespace channels, we should have all APIs to work with it to avoid converting from one to another

### CODE REVIEW

File: src/metabase/api/channel.clj [hunk 1]
```diff
@@ -23,21 +24,27 @@
...
-    (let [result (channel/can-connect? (models.channel/keywordize-type type) details)]
+    (let [result (channel/can-connect? type details)]
       (when-not (true? result)
         (throw (ex-info "Unable to connect channel" (merge {:status-code 400} result)))))
     (catch Exception e
       (throw (ex-info "Unable to connect channel" (merge {:status-code 400} (ex-data e)))))))
 
+(def ^:private ChannelType
+  ;; TODO can we add a decoder that turn a string to a keyword?
+  (mu/with-api-error-message
```
Comment (qnkhuat) [inline src/metabase/api/channel.clj:35]: @escherize, is there a way to make this schema decode to `keyword` automatically?

Comment (crisptrutski) [reply src/metabase/api/channel.clj:35]: I think you want something like this (warning used an LLM)

```clojure
(def ChannelType
  [:and
   [:or
    [:string {:coerce keyword}]
    [:keyword]]
   [:fn (fn [x] (= (namespace x) "channel"))]])
```

Comment (crisptrutski) [reply src/metabase/api/channel.clj:35]: Seems it works with [small tweak](https://malli.io/?value=%22channel%2Fwhat%22&schema=%5B%3Aand%0A%20%5B%3Aor%0A%20%20%5B%3Astring%20%7B%3Acoerce%20keyword%7D%5D%0A%20%20%5B%3Akeyword%5D%5D%0A%20%5B%3Afn%20(fn%20%5Bx%5D%20(%3D%20(namespace%20(keyword%20x))%20%22channel%22))%5D%5D)

File: src/metabase/api/channel.clj [hunk 2]
```diff
@@ -63,14 +70,16 @@
         type-changed?    (some-> type keyword (not= (:type channel-before-update)))]
 
     (when (or details-changed? type-changed?)
-      (test-channel-connection! (or type (:type channel-before-update))
+      (test-channel-connection! (or (keyword type) (:type channel-before-update))
```
Comment (crisptrutski) [inline src/metabase/api/channel.clj:73]: TIL you can nil pun with `keyword`

**Review by crisptrutski — COMMENTED** [review]
Malli not (yet) doing the coercion for us doesn't seem like  blocker.

Just sanity checking - we don't need a migration, because users haven't been able to create any channels yet? What about stats, is that still clean, or will we clean it up manually?

**Comment by qnkhuat** [comment]
good call out! all http related works are going in a feature branch `feat-http-channel` so we don't need a migration here.

File: src/metabase/api/channel.clj (hunk 1 above)
Comment (escherize) [reply src/metabase/api/channel.clj:35]: You could do it that way, or you could just put the coercion data onto the fn form itself. I'm not exactly sure what shapes you want to allow, but the way I'd mess with it in the repl is:

``` clojure

#_:clj-kondo/ignore ;;nocommit
(require '[malli.core :as mc] '[malli.error :as me] '[malli.util :as mut] '[metabase.util.malli :as mu]
         '[metabase.util.malli.describe :as umd] '[malli.provider :as mp] '[malli.generator :as mg]
         '[malli.transform :as mtx] '[metabase.api.common.internal :refer [defendpoint-transformer]])

(def my-schema [:fn {:decode/string (fn [x] (keyword x))}
                #(= "channel" (namespace (keyword %)))])

(mc/decode my-schema "channel/what" defendpoint-transformer)
;; => :channel/what


(map (fn [v] [v (mc/validate [:fn {:decode/string (fn [x] (keyword x))}
                             #(= "channel" (namespace (keyword %)))]
                            v)])
     ["channel"
      "channel/what"
      "channel/what/else"
      :channel
      :channel/who])
;; => (["channel" false]
;;     ["channel/what" true]
;;     ["channel/what/else" true]
;;     [:channel false]
;;     [:channel/who true])

```

Comment (qnkhuat) [reply src/metabase/api/channel.clj:35]: Isn's this lovely, thank you gentlemen! 6df536ff49

File: src/metabase/models/channel.clj [hunk 3]
```diff
@@ -28,10 +28,18 @@
...
-  (if (and (keyword? channel-type)
-           (= "channel" (namespace channel-type)))
-    channel-type
-    (keyword "channel" (name channel-type))))
+(defn- assert-channel-type
+  [{channel-type :type}]
+  (when-not (= "channel" (-> channel-type keyword namespace))
+    (throw (ex-info "Channel type must be a namespaced keyword like :channel/http" {:status-code  400
+                                                                                    :channel-type channel-type}))))
+
```
Comment (crisptrutski) [inline src/metabase/models/channel.clj:36]: Extra blank line here

File: test/metabase/models/channel_test.clj [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+
+(comment
+ ;; to register the :metabase-test channel implementation
+ api.channel-test/keepme)
+
+(deftest channel-details-is-encrypted
+  (encryption-test/with-secret-key "secret"
+   (mt/with-model-cleanup [:model/Channel]
+     (let [channel (t2/insert-returning-instance! :model/Channel {:name    "Test channel"
+                                                                  :type    "channel/metabase-test"
```
Comment (crisptrutski) [inline test/metabase/models/channel_test.clj:18]: Lots of extra lime breaks

**Review by crisptrutski — APPROVED** [review]

File: src/metabase/models/channel.clj (hunk 3 above)
Comment (qnkhuat) [reply src/metabase/models/channel.clj:36]: there is only one newline on 36

File: test/metabase/models/channel_test.clj (hunk 4 above)
Comment (qnkhuat) [reply test/metabase/models/channel_test.clj:18]: 🍋? :D 

I don't see any line breaks tho.

Comment (crisptrutski) [reply test/metabase/models/channel_test.clj:18]: Blame my phone screen 😄

File: src/metabase/models/channel.clj (hunk 3 above)
Comment (crisptrutski) [reply src/metabase/models/channel.clj:36]: Ditto not seeing the rest of the map on my phone 🤦

**Pull request merged**